zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -270,69 +270,156 @@ impl DbFormatChange {
+                        batch.delete_sapling_tree(&db, &height);
+                        db.write_batch(batch)
+                            .expect("Deleting note commitment trees should always succeed.");
                     }
 
-                    prev_orchard_tree = orchard_tree;
-                    prev_sapling_tree = sapling_tree;
+                    prev_height = height;
 
-                    height = height.next();
+                    warn!(?height, "Database upgrade is at:");
+                }
+            });
+
+            // Create an unbounded channel for reading note commitment trees
+            let (orchard_tree_tx, orchard_tree_rx) = mpsc::channel();
+
+            // Set up task for reading orchard note commitment trees
+            let db = upgrade_db.clone();
+            let orchard_read_task = std::thread::spawn(move || {
+                for (height, tree) in
+                    db.orchard_tree_by_height_range(orchard_height..initial_tip_height)
+                {
+                    let _ = orchard_tree_tx.send((height, tree));
+                }
+            });
+
+            // Create an unbounded channel for duplicate orchard note commitment tree heights
+            let (unique_orchard_tree_height_tx, unique_orchard_tree_height_rx) = mpsc::channel();
+
+            // Set up task for reading orchard note commitment trees
+            let orchard_compare_task = std::thread::spawn(move || {
+                while let Ok((height, tree)) = orchard_tree_rx.recv() {
+                    let tree = Some(tree);
+                    if prev_orchard_tree != tree {
+                        let _ = unique_orchard_tree_height_tx.send(height);
+                        prev_orchard_tree = tree;
+                    }
                 }
+            });
+
+            // Set up task for deleting orchard note commitment trees
+            let db = upgrade_db;
+            let orchard_delete_task = std::thread::spawn(move || {
+                let mut prev_height = orchard_height;
+                while let Ok(height) = unique_orchard_tree_height_rx.recv() {
+                    let delete_from = (prev_height + 1).expect("should be valid height");
+                    let delete_to = (height - 1).expect("should be a valid height");
+                    let num_entries = delete_to.0.checked_sub(delete_from.0);
+
+                    if num_entries.map_or(false, |n| n >= 1) {
+                        let mut batch: DiskWriteBatch = DiskWriteBatch::new();
+                        batch.delete_range_orchard_tree(&db, &delete_from, &delete_to);
+                        db.write_batch(batch)
+                            .expect("Deleting note commitment trees should always succeed.");
+                    } else if num_entries.map_or(false, |n| n == 0) {
+                        let mut batch: DiskWriteBatch = DiskWriteBatch::new();
+                        batch.delete_orchard_tree(&db, &height);
```

Comment:
I think this should be `height -1`. However, I think we can remove this `else if` and use only `delete_range_orchard_tree` since it doesn't delete the upper bound: https://docs.rs/rocksdb/latest/rocksdb/struct.WriteBatchWithTransaction.html#method.delete_range_cf?

Reply:
Fixed in https://github.com/ZcashFoundation/zebra/pull/7321/commits/acce514c7057b06985a8ef1df97b887e20d256f5.